### PR TITLE
SITES-950: chmod to overcome AFS ACLs

### PR DESCRIPTION
**Pull request by jbickar**
# READY FOR REVIEW

# Summary
- Fix file perms

# Needed By (Date)
- Before we migrate the next site with file permissions issues

# Criticality
- How critical is this PR on a 1-10 scale? 3/10

# Steps to Test

0. Check out `master`
1. `chmod 600 /afs/ir/dist/drupal/ds_sws-migration-testing/files/california-128.png`
2. Sync `sws-migration-testing` to dev
3. Go to https://sws-migration-testing-dev.sites.stanford.edu/file-permissions-test-sites-950
4. See no image
5. Check out this branch
6. Add `chmod: "TRUE"` to your `migration_vars.yml`
7. `chmod 600 /afs/ir/dist/drupal/ds_sws-migration-testing/files/california-128.png`
8. Sync `sws-migration-testing` to dev
9. Go to https://sws-migration-testing-dev.sites.stanford.edu/file-permissions-test-sites-950
10. See image of California flag

# Additional Contextual Information
1. This requires **both** `afs_available="TRUE"` _and_ a new variable, `chmod="TRUE"`
2. I did not see the sense in multi-purposing this for when AFS is not available; we only will need it in a very small set of circumstances.
3. This makes permanent (but reversible) changes to file permissions on the source site. It does not track the previous file permissions. This is a calculated risk between expediency and safety.
4. This is not an option that we want to use unless we have to. (It's not really that risky, but I don't want to get in the habit of making permanent file permission changes.)

# Affected Projects or Products
- Sites with images/files that have been manually synced

# Associated Issues and/or People
- JIRA ticket: https://stanfordits.atlassian.net/browse/SITES-950
- Anyone else who should be notified? @pookmish and @kbrownell 

# See Also
- [PR Checklist](https://gist.github.com/sherakama/0ba17601381e3adbe0cad566ad4d80a5)

### CODE REVIEW

File: roles/post-db-restore/tasks/main.yml [hunk 1]
```diff
@@ -99,6 +99,16 @@
   when: launch_tasks == ""
   notify: Clear site cache
 
+# chmod files directory so that there are not permissions issues when going from
+# AFS ACLs to Unix file permissions
+- name: Make all files readable on AFS
+  file:
+    path: "/afs/ir/dist/drupal/{{ site_prefix }}_{{ inventory_hostname }}/files/"
+    recurse: yes
+    mode: "a+r"
```
Comment (jbickar) [inline roles/post-db-restore/tasks/main.yml]: This is acceptable because AFS ignores Unix file permissions. Thus they are served up by Sites 1.x despite being 600 and owned by `jbickar` (e.g.).

An argument could be made to make this `u+r,g+r`; I would entertain such an argument.

Note that this also affects private files on AFS (they live in `/afs/ir/dist/drupal/{{ site_prefix }}_{{ inventory_hostname }}/files/private/`. Those are protected by `.htaccess` rules. You can verify this by trying to access https://sws-oddball-vhost.stanford.edu/sites/default/files/private/stanford_page/about_stacks.pdf.

```
jbickar@people1:/afs/ir/dist/drupal/ds_sws-migration-testing/files$ ll private/stanford_page/about_stacks.pdf 
-rw-rw-r-- 1 80981 users 456K Aug 31 07:54 private/stanford_page/about_stacks.pdf
```

Even when it's 664, https://sws-oddball-vhost.stanford.edu/sites/default/files/private/stanford_page/about_stacks.pdf gives a 403.

Comment (sherakama) [reply roles/post-db-restore/tasks/main.yml]: I found this thread: https://www.drupal.org/forum/support/post-installation/2016-09-22/file-and-directory-permissions-lets-finally-get-this

TLDR; Directories should be 0755 and files should be 0644.

Comment (sherakama) [reply roles/post-db-restore/tasks/main.yml]: I can agree with that approach as I have had to make directories 0755 for Drupal to be able to create and modify nested files and directories.

Comment (jbickar) [reply roles/post-db-restore/tasks/main.yml]: `a+r` is more conservative than `755`/`644`. However, directories **do** need to have the executable bit set. Hm.

Stack Overflow to the rescue: https://stackoverflow.com/questions/28778738/ansible-mode-755-for-directories-and-644-for-files-recursively

TL;DR:
```
mode: u=rwX,g=rX,o=rX
```

**Comment by jbickar** [comment]
@sherakama, ready for another pass. Testing steps should be the same.

File: group_vars/all.yml [hunk 2]
```diff
@@ -44,3 +44,7 @@ enable_maintenance_mode_source: "FALSE"
 # Set disable_maintenance_mode_dest to TRUE initially, to avoid undefined
 # variable errors. It can be overridden in migration_vars.yml
 disable_maintenance_mode_dest: "FALSE"
+
+# This is a little-used variable; only used when there are file permissions
+# issues from files that were not created/saved by Drupal.
+chmod: "FALSE"
```
Comment (sherakama) [inline group_vars/all.yml]: I'd say we would always want to do this. It doesn't hurt and would save us having to discover which sites have this issue.

File: roles/post-db-restore/tasks/main.yml [hunk 3]
```diff
@@ -99,6 +99,16 @@
...
 
+# chmod files directory so that there are not permissions issues when going from
+# AFS ACLs to Unix file permissions
+- name: Make all files readable on AFS
+  file:
+    path: "/afs/ir/dist/drupal/{{ site_prefix }}_{{ inventory_hostname }}/files/"
+    recurse: yes
+    mode: "u=rwX,g=rX,o=rX"
+  when:
+    afs_available == "TRUE" and chmod == "TRUE"
```
Comment (sherakama) [inline roles/post-db-restore/tasks/main.yml]: Shouldn't we also do this on the local version?

Comment (sherakama) [reply roles/post-db-restore/tasks/main.yml]: Or are we not running this on local anymore and should assume my-vbox or the people servers?

File: group_vars/all.yml (hunk 2 above)
Comment (jbickar) [reply group_vars/all.yml]: Hm. I wonder how much time it takes. However, we're done with the mass migrations, where seconds count, so it's probably worth it here.

File: roles/post-db-restore/tasks/main.yml (hunk 3 above)
Comment (jbickar) [reply roles/post-db-restore/tasks/main.yml]: Yeah, sure, wouldn't hurt on local too. Although I haven't run on local (my-stanford-vbox, in my case) other than for limited testing since we got people1/people2 set up. So much faster on those two boxes.

I got Bill set up on people1/2 and am steering him in that direction, so I'm kinda "meh" on this.
